### DISCUSSION TITLE
Rego package ref-arg-brack

### DISCUSSION THREAD

**Discussion by Ronnie-personal** [open-policy-agent/community#485] [OPA and Rego] [answered]
Hi,

I'm reading the document at https://www.openpolicyagent.org/docs/latest/policy-reference/#grammar. 

package         = "package" ref

ref             = ( var | array | object | set | array-compr | object-compr | set-compr | expr-call ) { ref-arg }
ref-arg         = ref-arg-dot | ref-arg-brack
ref-arg-brack   = "[" ( scalar | var | array | object | set | "_" ) "]"

Can you please help me understand the use cases for ref-arg-brack in the package? What problem does this feature solve?

Thank you!

**Comment by anderseknert** [comment]
Hi Ronnie! An example would be when you want or need to include characters which aren’t allowed in vars, like dashes.

`package rules.i-am-dashing`

doesn’t work, but

`package rules[“i-am-dashing”]`

does

**Reply by Ronnie-personal** [reply]
Thanks for the prompt reply.
So when I call OPA Rest API to evaluate policy, the URL path can be /v1/data/rules/i-am-dashing/, correct? And I can have multiple ref-arg-brack, package rules["arg-1"]["arg-2"]?

Is this grammar 'ref-arg-brack = "[" ( scalar | var | array | object | set | "_" ) "]"' accurate? I don't see the 'dash'.

**Reply by srenatus** [reply, marked as answer]
I'm afraid the grammar is most likely wrong. For one thing, `"package" ref-arg-brack` doesn't describe only _valid_ package lines. The following would be valid-per-grammar but invalid for rego:

```rego
package foo[_][x]
```

> And I can have multiple ref-arg-brack, package rules["arg-1"]["arg-2"]?

Yes. You could also use brackets where they're unnecessary, like `package foo["bar"]`, I think.

> Is this grammar 'ref-arg-brack = "[" ( scalar | var | array | object | set | "_" ) "]"' accurate? I don't see the 'dash'.

It should be covered by scalar? Since `"i-am-dashing"` is a string?

Also, as another word of warning, the grammar in the docs is a bit... non-optimal insofar as it's just text in a document. It's not an artifact that would be tied into our parser code, or used for validation, or anything. So it looses touch with reality quite easily.
